utils/logger.go

```diff
@@ -0,0 +1,129 @@
+package utils
+
+import (
+	"strings"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// Logger is a wrapper for zap.SugaredLogger.
+type Logger struct {
+	zLogger *zap.SugaredLogger
+}
+
+// A LoggerConfig contains the running environment
+// which is either "development" or "production",
+// and the path of file to write the logging output to.
+// Logging is enabled at DebugLevel and above in development mode,
+// at InfoLevel and above in production mode.
+type LoggerConfig struct {
+	EnableStacktrace bool   `toml:"enable_stacktrace,omitempty"`
+	Environment      string `toml:"env"`
+	Path             string `toml:"path"`
+}
+
+// NewLogger creates an instance of Logger with
+// default configurations, i.e., the logger writes
+// DebugLevel and above logs in development environment,
+// InfoLevel and above logs in production environment
+// to a JSON-format file.
+func NewLogger(conf *LoggerConfig) *Logger {
+	zLevel := zap.NewAtomicLevel()
+	switch {
+	case strings.EqualFold("development", conf.Environment):
+		zLevel.SetLevel(zap.DebugLevel)
+	case strings.EqualFold("production", conf.Environment):
+		zLevel.SetLevel(zap.InfoLevel)
+	default:
+		panic("Environment must be either development or production")
+	}
+
+	zConfig := &zap.Config{
+		Level:             zLevel,
+		Development:       strings.EqualFold("development", conf.Environment),
+		Encoding:          "json",
+		DisableStacktrace: !conf.EnableStacktrace, // the developer needs to explicitly enable this
+		EncoderConfig: zapcore.EncoderConfig{
+			TimeKey:        "timestamp",
+			LevelKey:       "level",
+			NameKey:        "logger",
+			CallerKey:      "path",
+			MessageKey:     "msg",
+			StacktraceKey:  "stack",
+			EncodeLevel:    zapcore.CapitalLevelEncoder,
+			EncodeTime:     zapcore.ISO8601TimeEncoder,
+			EncodeDuration: zapcore.StringDurationEncoder,
+		},
+		OutputPaths: []string{conf.Path},
+	}
+
+	logger, err := zConfig.Build()
+	if err != nil {
+		panic(err)
+	}
+	return &Logger{logger.Sugar()}
+}
+
+// Debug logs a message that is most useful to debug,
+// with some additional context addressed by key-value pairs.
+func (l *Logger) Debug(msg string, keysAndValues ...interface{}) {
```

Comment:
The signature helps the developers know that these could be used like:
`l.Debug(fmt.Sprintf())` or `l.Debug(msg, key, value)`, but not `l.Debug(msg1, msg2)`.